types/react-router/index.d.ts

```diff
@@ -129,7 +130,7 @@ export type Omit<T, K extends keyof T> = T extends any ? Pick<T, Exclude<keyof T
 
 export function matchPath<Params extends { [K in keyof Params]?: string }>(pathname: string, props: string | RouteProps, parent?: match<Params> | null): match<Params> | null;
 
-export function generatePath(pattern: string, params?: { [paramName: string]: string | number | boolean | undefined }): string;
+export function generatePath<Params extends { [K in keyof Params]?: string }>(pattern: string, params?: Params): string;
```

Comment:
We could use `Params extends Record<string, string | number | boolean | undefined>` which is more obvious and restores the previous value type. The current typing has this circular reference i.e. Params extends keyof Params which is always harder to reason about.